src/storage/src/source/types.rs

```diff
@@ -189,3 +196,31 @@ pub struct DecodeResult<FromTime> {
     /// The original timestamp of this message
     pub from_time: FromTime,
 }
+
+#[pin_project]
+pub struct SignaledFuture<F> {
+    #[pin]
+    fut: F,
+    semaphore: PollSemaphore,
+}
+
+impl<F: Future> SignaledFuture<F> {
+    pub fn new(semaphore: Arc<Semaphore>, fut: F) -> Self {
+        Self {
+            fut,
+            semaphore: PollSemaphore::new(semaphore),
+        }
+    }
+}
+
+impl<F: Future> Future for SignaledFuture<F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let permit = ready!(this.semaphore.poll_acquire(cx));
+        let ret = this.fut.poll(cx);
+        drop(permit);
```

Comment:
is this here just to make it obvious that we are dropping this semaphore permit? or is there some other thing that occurs by dropping explicitly rather than implicitly when the method returns?

Reply:
No, just making it explicit so that if its creation is removed it will trigger a compile error